src/custom_de.rs

```diff
@@ -94,31 +104,39 @@ impl<'de> de::Deserializer<'de> for &mut EthFixedDeserializer {
     }
 
     fn deserialize_u32<V: de::Visitor<'de>>(self, visitor: V) -> Result<V::Value> {
+        if self.remaining_bytes <= 0 {
+            panic!("received u8 when there are no more expected bytes to deserialize")
+        }
+
         match self.serializer_type {
             Fixed::H256 | Fixed::H160 => panic!("received u32 when deserializing H256,H160"),
             Fixed::U256 => {
-                let value = (self.content[self.offset as usize] as u32)
-                    + ((self.content[(self.offset + self.offset_sign) as usize] as u32) << 8)
-                    + ((self.content[(self.offset + 2 * self.offset_sign) as usize] as u32) << 16)
-                    + ((self.content[(self.offset + 3 * self.offset_sign) as usize] as u32) << 24);
+                self.remaining_bytes -= 4;
+                let mut value = 0u32;
+                for byte_index in 0..4 {
+                    let index = (self.offset + byte_index * self.offset_sign) as usize;
+                    value += (self.content[index] as u32) << (8 * byte_index);
+                }
                 self.offset += 4 * self.offset_sign;
                 visitor.visit_u32(value)
             }
         }
     }
 
     fn deserialize_u64<V: de::Visitor<'de>>(self, visitor: V) -> Result<V::Value> {
+        if self.remaining_bytes <= 0 {
```

Comment:
I don't know what is idiomatic rust, but in C++ templates I'd be doing the equivalent of `if self.remaining_bytes < sizeof(T) {` where T is `u64` in this case, and on L134 the code would be the moral equivalent of `self.remaining_bytes -= sizeof(T)`.  It is good to use the same expression (here, a magic constant) in the matching lines.

Reply:
@bennetyee yes I see the point. There's just a few places where we need to do this, and it doesn't look like we will need to add more checks for this, so if that's fine with you we can keep it the same.

Otherwise we can write generic function to deserialize the bytes for any integer type provided, but I'm not sure it's worth the complexity.

Reply:
It is ok to just correct the magic numbers -- it's borderline size-/code replication-wise in terms of the trade-off between replicated verification/review (code replication lulls reviewers into skimming and missing errors) versus slightly more complex generic code which can be reviewed/audited once.  I _prefer_ less code overall that the generics approach would give us, and to have what code that remain be more obviously correct code w/o magic constants, but understand it's a lot more effort than fixing the magic numbers.  But if we add more integer types (even more replication), we should _definitely_ switch to generics.

Reply:
agreed, thanks @bennetyee